File: packages/smart-reply/ios/Plugin/SmartReplyPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(SmartReplyPlugin)
public class SmartReplyPlugin: CAPPlugin {
    public static let tag = "SmartReply"

    private var implementation: SmartReply?

    override public func load() {
        implementation = SmartReply()
    }

    @objc func suggestReplies(_ call: CAPPluginCall) {
        do {
            let options = try SuggestRepliesOptions(call)

            implementation?.suggestReplies(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, result)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", SmartReplyPlugin.tag, "] ", error)
        call.reject(error.localizedDescription)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/smart-reply/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/smartreply/SmartReplyPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.smartreply;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.classes.options.SuggestRepliesOptions;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.classes.results.SuggestRepliesResult;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.mlkit.smartreply.interfaces.Result;

@CapacitorPlugin(name = "SmartReply")
public class SmartReplyPlugin extends Plugin {

    public static final String TAG = "SmartReply";

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error occurred.";

    private SmartReply implementation;

    @Override
    public void load() {
        implementation = new SmartReply();
    }

    @Override
    protected void handleOnDestroy() {
        if (implementation != null) {
            implementation.close();
        }
    }

    @PluginMethod
    public void suggestReplies(PluginCall call) {
        try {
            SuggestRepliesOptions options = new SuggestRepliesOptions(call);
            NonEmptyResultCallback<SuggestRepliesResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull SuggestRepliesResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.suggestReplies(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
